src/freertos_drivers/esp_idf/EspIdfWiFi.cxx

```diff
@@ -0,0 +1,1375 @@
+            {
+                // First reconnect attempt, or forced "fast" connect.
+                esp_wifi_connect();
+                LOG(INFO, "wifi: STA disconnected, fast reconnect attempt...");
+            }
+            else
+            {
+                // Successive reconnect attempts.
+                esp_wifi_scan_start(nullptr, false);
+                LOG(INFO, "wifi: STA disconnected, scanning...");
+            }
+            std::string ssid((char*)evdata->ssid, evdata->ssid_len);
+            wlan_connected(
+                false, connection_result_encode(evdata->reason), ssid);
+            break;
+        }
+        case WIFI_EVENT_SCAN_DONE:
+        {
+            /// @todo The newer IDF versions have a new api that can read
+            ///       AP records one at a time, which is much more memory
+            ///       friendly. Revisit this when updating the IDF version.
+            int idx = -1;
+            uint16_t number;
+            esp_wifi_scan_get_ap_num(&number);
+            LOG(INFO, "wifi: Scan done, number of records: %u.", number);
+            number = std::min(number, static_cast<uint16_t>(10));
+
+            // This takes a lot of memory. Not the best API design. See note
+            // above.
+            wifi_ap_record_t *ap_records = new wifi_ap_record_t[number];
+            esp_wifi_scan_get_ap_records(&number, ap_records);
+            std::string ssid(MAX_SSID_SIZE + 1, '\0');
+            std::string pass(MAX_PASS_SIZE + 1, '\0');
+            uint8_t sec = WIFI_AUTH_WPA2_PSK;
+            uint8_t conn_channel = 0;
+            {
+                OSMutexLock locker(&lock_);
+                scanResults_.clear();
+                for (unsigned i = 0; i < number; ++i)
+                {
+                    // Cache the scan record for the user to read later.
+                    scanResults_.emplace_back((char*)ap_records[i].ssid,
+                        sec_type_encode(ap_records[i].authmode),
+                        ap_records[i].rssi);
+
+                    if (staIface_ && !connected_ && idx < 0)
+                    {
+                        int index = 0;
+                        do
+                        {
+                            std::string profile_pass;
+                            uint8_t profile_sec;
+                            index = find_sta_profile(
+                                std::string((char*)ap_records[i].ssid),
+                                &profile_pass, &profile_sec, index);
+                            if (index >= 0 &&
+                                ap_records[i].authmode >= profile_sec)
+                            {
+                                // profile SSID and security mode match.
+                                idx = index;
+                                ssid = (char*)ap_records[i].ssid;
+                                pass = std::move(profile_pass);
+                                sec = profile_sec;
```

Comment:
this should be the ap authmode, not the profile authmode

Reply:
I will work either way, but there are some corner cases where ap authmode is better, so I changed it.